art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py

```diff
@@ -0,0 +1,450 @@
+        :type eps_step: `float`
+        :param random_init: Random initialisation within the epsilon ball. For random_init=False
+            starting at the original input.
+        :type random_init: `bool`
+        :return: Adversarial examples.
+        :rtype: `Tensor`
+        """
+        if random_init:
+            n = x.shape[0]
+            m = np.prod(x.shape[1:])
+            x_adv = x.astype(ART_NUMPY_DTYPE) + (
+                random_sphere(n, m, eps, self.norm).reshape(x.shape).astype(ART_NUMPY_DTYPE)
+            )
+
+            if hasattr(self.estimator, "clip_values") and self.estimator.clip_values is not None:
+                clip_min, clip_max = self.estimator.clip_values
+                x_adv = np.clip(x_adv, clip_min, clip_max)
+            x_adv = torch.from_numpy(x_adv).to(self.estimator.get_device)
+
+        else:
+            if isinstance(x, np.ndarray):
+                x_adv = torch.from_numpy(x.astype(ART_NUMPY_DTYPE)).to(self.estimator.get_device)
+            else:
+                x_adv = x
+
+        # Get perturbation
+        perturbation = self._compute_perturbation(x_adv, y, mask)
+
+        # Apply perturbation and clip
+        x_adv = self._apply_perturbation(x_adv, perturbation, eps_step)
+
+        # Do projection
+        perturbation = self._projection(x_adv - x_init, eps, self.norm)
+
+        # Recompute x_adv
+        x_adv = perturbation + x_init
+
+        return x_adv
+
+    @staticmethod
+    def _projection(values, eps, norm_p):
+        """
+        Project `values` on the L_p norm ball of size `eps`.
+
+        :param values: Values to clip.
+        :type values: `Tensor`
+        :param eps: Maximum norm allowed.
+        :type eps: `float`
+        :param norm_p: L_p norm to use for clipping. Only 1, 2 and `np.Inf` supported for now.
+        :type norm_p: `int`
+        :return: Values of `values` after projection.
+        :rtype: `Tensor`
+        """
+        # Pick a small scalar to avoid division by 0
+        tol = 10e-8
+        values_tmp = values.reshape(values.shape[0], -1)
+
+        if norm_p == 2:
+            values_tmp = values_tmp * torch.min(
+                torch.FloatTensor([1.0]), eps / (torch.norm(values_tmp, p=2, dim=1) + tol)
+            ).unsqueeze_(-1)
+
+        elif norm_p == 1:
+            values_tmp = values_tmp * torch.min(
+                torch.FloatTensor([1.0]), eps / (torch.norm(values_tmp, p=1, dim=1) + tol)
+            ).unsqueeze_(-1)
+
+        elif norm_p == np.inf:
+            values_tmp = values_tmp.sign() * torch.min(values_tmp.abs(), torch.FloatTensor([eps]))
+
+        else:
+            raise NotImplementedError(
+                "Values of `norm_p` different from 1, 2 and `np.inf` are currently not supported."
+            )
+
+        values = values_tmp.reshape(values.shape)
+
+        return values
+
+    def set_params(self, **kwargs):
+        """
```

Comment:
It could be that the `set_params` methods of all ProjectedGradientDescent* are the same or very similar. Would it be possible to define it in`ProjectedGradientDescent` and reuse it in `ProjectedGradientDescentNumpy`, `ProjectedGradientDescentTensorFlow` and `ProjectedGradientDescentPyTorch`?

Reply:
No, we cannot do that because there are no inheritance relationship between the main PGD and different implementations of it.

Reply:
Might it be worthwhile then to create a new parent class for `ProjectedGradientDescentNumpy`, `ProjectedGradientDescentTensorFlow` and `ProjectedGradientDescentPyTorch` to have the common logic (in particular from `__init__` and `set_params`, as well as the checks at the beginning of `generate`) in one place?

Reply:
I have created a new parent class to include this comment.